# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [5.0.14] - 2025-06-20
### Run Processing (Illumina)
- adding Illumina Single Cell UDI 96 rxn indexes.
- adding Illumina Single Cell UDI mix strip indexes.
- adding Singleron GEXSCOPE scRNA indexes.
### Sample Manifest
- same as above.

###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.14%0D5.0.13#diff)




## [5.0.13] - 2025-03-14
### Run Processing (Illumina)
- adding Parse Bio sublibrary indexes 3-16.
### Sample Manifest
- adding Parse Bio sublibrary indexes 3-16.

###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.13%0D5.0.12#diff)





## [5.0.12] - 2025-02-26
### Run Processing (Illumina)
- added Cantata Bio Dovetail indexes.
### Sample Manifest
- added Cantata Bio Dovetail indexes.

###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.12%0D5.0.11#diff)


## [5.0.11] - 2025-01-15
### Run Processing (Illumina)
- removed blast step for Amplicon library types (Amplicon DNA, Amplicon RNA, 16S, 18S)

###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.11%0D5.0.10#diff)


## [5.0.10] - 2024-12-20
### Run Processing (Illumina)
- added index Bailey4F.
### Sample Manifest
- added index Bailey4F.

###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.10%0D5.0.9#diff)


## [5.0.9] - 2024-12-11
### Run Processing (Illumina)
- Decreased Wall time for blast step from 96h to 24h.

###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.9%0D5.0.8#diff)




## [5.0.8] - 2024-11-12
### Run Processing (MGI)
- Added MGIeasy Fast FS index definitions.
### Sample Manifest
- Added MGIeasy Fast FS index definitions.
- Removed Homo_sapiens:GRCh37 and Homo_sapiens:hg19 from reference genome drop-down menu.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.8%0D5.0.7#diff)



## [5.0.7] - 2024-10-24
### LIMS Configuration
- __library step__ adding IDT Illumina-Nextera V3 indexes to PCR-Free library step (added manually in prod 2024-10-06)
### Run Processing (Illumina)
- Adding string "V3" to all IDT Illumina-Nextera V3 index names.
### Sample Manifest
- Adding string "V3" to all IDT Illumina-Nextera V3 index names.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.7%0D5.0.6#diff)




## [5.0.6] - 2024-09-23
### Run Processing (Illumina)
- Adding IDT Illumina-Nextera V3 indexes.
### Sample Manifest
- Adding IDT Illumina-Nextera V3 indexes.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.6%0D5.0.5#diff)



## [5.0.5] - 2024-08-23
### Run Processing (Illumina)
- Adding GeoMx ProCode Index series.
### Sample Manifest
- Adding GeoMx ProCode Index series.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.5%0D5.0.4#diff)



## [5.0.4] - 2024-08-07
### Run Processing (Illumina)
- Adding custom indexes for Nano Cut and Tag.
### Sample Manifest
- Adding custom indexes for Nano Cut and Tag.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.4%0D5.0.3#diff)



## [5.0.3] - 2024-07-16
### Run Processing (Illumina)
- Adding indexes for Fluent BioSciences - PIPseq T20 3 Prime Single Cell RNA Kit.
### Sample Manifest
- Adding indexes for Fluent BioSciences - PIPseq T20 3 Prime Single Cell RNA Kit.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.3%0D5.0.2#diff)

## [5.0.2] - 2024-05-31
### Run Processing (Illumina)
- Correcting sequence for Bailey4R index.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.2%0D5.0.1#diff)


## [5.0.1] - 2024-05-08
### Run Processing (Illumina)
- Adding Custom indexes for Bailey Lab.
### Sample Manifest
- See above.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.1%0D5.0.0#diff)


## [5.0.0] - 2024-04-19
### Run Processing (Illumina)
- Adding NovaSeqX support.
### Run Monitor (automatic launcher)
- Adding NovaSeqX support.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/5.0.0%0D4.0.19#diff)

## [4.0.19] - 2024-03-28
### Run Processing (Illumina)
- Adding New England Biolabs (NEB) - NEBNext Enzymatic Methyl-seq 96 Reactions (NEB #E7140L) indexes
### Sample Manifest
- see above
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.19%0D4.0.18#diff)

## [4.0.18] - 2024-03-22
### Run Processing (Illumina)
- Adding Honeycomb - HIVE scRNAseq UDI indexes (Uses custom sequencing primer different from v1 kit)
### Sample Manifest
- see above
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.18%0D4.0.17#diff)

## [4.0.17] - 2024-01-17
### Run Processing (Illumina)
- Correcting index for Honeycomb - HIVE scRNAseq v1 Kit
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.17%0D4.0.16#diff)

## [4.0.16] - 2023-12-15
### Run Processing (Illumina)
- Adding Honeycomb - HIVE scRNAseq v1 Kit (Uses custom sequencing primer)
- Adding scRNA-Seq library type
### Sample Manifest
- see above
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.16%0D4.0.15#diff)

## [4.0.14] - 2023-12-01
### Run Processing (Illumina)
- Adding Bimodal - Duet Multiomics ModC UDI adapters
### Sample Manifest
- see above
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.14%0D4.0.13#diff)

## [4.0.13] - 2023-11-24
### Run Processing (Illumina)
- Correction to Diagenode index Diag_Set3_UDI_81_i5 from TCTCATTC to ACGCCGCA
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.13%0D4.0.12#diff)

## [4.0.12] - 2023-11-16
### Run Processing (Illumina)
- Adding quantabio indexes.
### Sample Manifest
- Adding quantabio indexes.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.12%0D4.0.11#diff)

## [4.0.11] - 2023-10-31
### Run Processing (Illumina)
- Adding new cut and tag index.
### Sample Manifest
- Adding new cut and tag index.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.11%0D4.0.10#diff)

## [4.0.10] - 2023-09-21
### Run Processing (MGI G400)
- Added multiQC command at the end of the clarity-genpipes pipeline call to complete the processing.
### LIMS Configuration
- Adding PacBio sequencing steps to the protocol list for PacBio event file generation.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.10%0D4.0.9#diff)

## [4.0.9] - 2023-08-02  
### Run Processing (Illumina)
- Correction to Integrated DNA Technologies (IDT) - xGen UDI-UMI Methylated Adapters order 18636890
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.9%0D4.0.8#diff)

## [4.0.8] - 2023-07-11  
### Run Processing (Illumina)
- Adding new index for Twist biosciences - Twist Universal Adapter System 16 UDI primers.
- Adding Twist Biosciences - Twist Universal Adapter System UDI Plates A/B/C/D
### Sample Manifest
- Adding new index for Twist biosciences - Twist Universal Adapter System 16 UDI primers. 
- Adding Twist Biosciences - Twist Universal Adapter System UDI Plates A/B/C/D
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.8%0D4.0.7#diff)

## [4.0.7] - 2023-06-27  
### Run Processing (Illumina)
- Adding new index for BD biosciences - BD rhapsody Whole Transciptome.
- Adding new index for Roche KAPA Unique Dual Index.
- Changed bwa version from v0.7.10 to v0.7.17
- Removed printf statement from -R option in bwa mem command.
- Added bwa mem options: "-Y -K 100000000"
- Removed bwa mem option: -M
- Changed centre from "McGill University and Genome Quebec Innovation Centre" to "McGill Genome Centre"
### Sample Manifest
- Adding new index for D biosciences BD rhapsody Whole Transciptome. 
- Adding new index for Roche KAPA Unique Dual Index.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.7%0D4.0.6#diff)

## [4.0.6] - 2023-05-17
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding library single cell ATAC multiome.
- __sample reception__, __all index validation__ - Adding index for molecularloop.
- __sample reception__, __all index validation__ - Adding index for WGBS index production (IDT new custom set) not published manifest.
### Run Processing Index Definitions
- Same as for lims configuration.
### Sample Manifest
- Adding library single cell ATAC multiome.
- Adding index for molecularloop.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.6%0D4.0.5#diff)

## [4.0.5] - 2023-04-05
### MGI Run Processing
- Fixed 'allrev' table update_merged_barcode_table.rb for MGI use
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.5%0D4.0.4#diff)

## [4.0.4] - 2023-04-03
### Run Processing
- Adding new index for 10x genomics TN series
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.4%0D4.0.3#diff)

## [4.0.3] - 2023-03-24
### Run Processing
- Adding new index for Parse Bio - Whole Transcriptome
- Correcting index QIASeq_miRNA_MIHT56
### Sample Manifest
- Adding new index for Parse Bio - Whole Transcriptome
- Correcting index QIASeq_miRNA_MIHT56
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.3%0D4.0.2#diff)

## [4.0.2] - 2023-02-23
### Run Processing
- Adding new index for DAIcrisper (no. 11, 12).
### Sample Manifest
- Adding new index for DAIcrisper (no. 11, 12).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.2%0D4.0.1#diff)

## [4.0.1] - 2023-02-22
### Run Processing
- Adding new index for parse bio.
- Adding new index for 10x genomics TS series.
### Sample Manifest
- Adding new index for parse bio.
- Adding new index for 10x genomics TS series.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.1%0D4.0.0#diff)

## [4.0.0] - 2022-12-12
### Nanopore covid run processing (event/lims trigger service)
- Added new Nanopore covid QC report to automatic run processing.
### Olink run processing
- Added new component to Olink run processing to run bcl2counts on Olink novaseq runs and email the result.
- Permanent folder created under processing/olink
### Secondary Analysis
- Added new Illumina covid QC report with automatic secondary analysis with genpipes covid pipeline for illumina.
- Permanent folder created under processing/secondary/illumcovidanalysis
### Run Processing (Illumina)
- Adding QIASeq Direct UDI Adapters definitions.
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding QIASeq Direct UDI Adapters definitions.
### Sample Manifest
- Correction to GenDX index pairing.
- Adding QIASeq Direct UDI Adapters to manifest.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/4.0.0%0D3.3.1#diff)

## [3.3.1] - 2022-11-29
### Run Processing
- Correcting Diagenode indexes from previous version.
- Adding new index Perkin Elmer - Bioo Scientific Nextflex Small RNA Adapters (8 plex kit)
### Sample Manifest
- Adding new index Perkin Elmer - Bioo Scientific Nextflex Small RNA Adapters (8 plex kit)
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/3.3.1%0D3.3.0#diff)

## [3.3.0] - 2022-11-16
### Illumina Run Processing
- Removing intel process flag (no longer necessary)
- Extending time for picard_rna_metrics (5h->10h) and rnaseq_qc (15h->24h) to accomodate very large RNASeq fastqs (>800M read pairs).
### Adding Custom Manifests to repo
- CentrEAU manifest to repo. Based on v3.0.0 version with added metadata columns after comment field.
- Ferri Group based on v3.2.0.
### Run Processing
- Adding Diagenode indexes (UDI Nextera 8bp) to index definition.
- Additional changes to run processing email lists.
### Sample Manifest
- Adding Diagenode indexes (UDI Nextera 8bp) to index definition.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/3.3.0%0D3.2.1#diff)

## [3.2.1] - 2022-10-17
### Run Processing (run monitor)
- Added force_mem_per_cpu option when calling ONT-covid pipeline.
- Adding Danielle Perley to run processing emails.
- Removing Sarah Reiling, Tony Tyr, Kaileigh Loranger and Rob Syme from run processing emails.
### Run Processing (Nanopore)
- Changes to genpipes in new commit genpipes/8279cfa from previous commit genpipes/4.2.0 covers issue with jobs being excluded from some nodes on abacus due to pmem flag.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/3.2.1%0D3.2.0#diff)

## [3.2.0] - 2022-09-20
### Illumina run processing
- Removed qc_graph tool to reduce runtime of all illumina processing (replaced by fastqc)
- Keeping bwa alignment only for these library types: PCR-free, PCR-enriched, RNASeq, 10x single cell, Panel, Amplicon DNA, AMplicon RNA, ChIPSeq, ChIPmentation and ATACSeq.
- Removing blasts for kapa tag spike-in verification (spike no longer used).
### MGI run processing
- Changed ruby script for countilluminabarcodes to also output an index sequence for the MGI pipelines.
### Olink alert Script
- Adding Olink Alert Script to repository. The script alerts the platform and Markus.
### Report script
- Added the script_clarity_extract.sh report script to the repo (used for CoVSeQ).
### Sample manifest
- Adding library types for GeoMx.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/3.2.0%0D3.1.2#diff)

## [3.1.2] - 2022-08-03
### LIMS Configuration
- Adding specific containers to NovaSeq (SP, S1, S2, S4), iSeq, MiSeq, G400 and T7 to reflect actual number of lanes available. (added manually in prod 2022-08-03)
- Adding shifted methylated IDT index set to lims in WGBS steps. (added manually in prod 2022-08-03)
### Illumina Run processing
- Removed duplicate HiC entry in library definitions. (added manually in prod 2022-08-01)
- Changes to email list for rapid reponse.
### Run Processing (Nanopore-covid)
- Changes to email list for rapid reponse.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/3.1.2%0D3.1.1#diff)

## [3.1.1] - 2022-07-12
### Illumina Run processing
- Adding SHARE-Seq (ATAC, RNA) handling in the pipeline to use Index 2 as primary index and Index 1 as an extra read.
- Removing cycle grouping from fastqc plots.
- Adding Ju-Ling Liu to illumina rapid response emails (covid related) and nanopore covid rapid response emails.
### Support Utilities
- New script to generate index set sequences using index definitions.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/3.1.1%0D3.1.0#diff)

## [3.1.0] - 2022-07-11
### LIMS Configuration
- __library step__ - Adding protocol for DNA Nanopore library (barcoding compatible).
- __library step__ - Adding protocol for RNA Nanopore library (barcoding compatible).
### Run Processing
- New nanopore protocols remain compatible with script generating event file for run processing.
- Generic nanopore run processing pipeline to be added.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/3.1.0%0D3.0.0#diff)

## [3.0.0] - 2022-06-29
### Support Utilities - New sample manifest update script (for new index and library type)
- For every new index definition and library type created in the lims (and marked to publish), this script will create the corresponding new manifest definitions for excel with no manual adjustments required in excel (No need to manually setup or update any dropdown ranges in excel).
- The script located here: runprocessing/generate_manifest_ranges.sh
- The procedure is described here: samplesubmissiontemplates/README.md
### Sample Manifest - Index kit and library type formating changes: 
- Index set names were updated to include the company in the name, if they are custom or not, single or dual. (script generated)
- Index set names now include index cycles (eg: 8bp-8bp, 10bp-10bp). (script generated)
### Sample Manifest - New index kits in dropdowns (script generated):
- 10x Genomics - Dual Index Kit NN Set A (eg: SI_NN7_A01-SI_NN5_A01)
- Agilent - SureSelect XT HS (high-sensitivity) (eg: SSXTHS701)
- Broad Institute - sgRNA (guide RNA) index definition (eg: BROAD_sgRNA_A01)
- Broad Institute - SmartSeq2 index (combinatorial dual) (eg: BRD701-BRD501|BRD502|...|BRD524)
- Dovetail - HiChIP MNase Kit (eg: DT_Set1_UDI_PrimerPair1_i7-DT_Set1_UDI_PrimerPair1_i5)
- GenDX -  NGSgo IndX plate I/II/III/IV (GenDX #2342456) (eg: GenDx001_I701-I501|I502|...|I516)
- Illumina - Made by IDT for Illumina-Nextera UDP (version 1) (eg: ILLUDP70001-ILLUDP50001)
- Illumina - Nextera DNA CD Indexes (combinatorial dual) (eg: H701-H501|H502|...|H522)
- Integrated DNA Technologies (IDT) - xGen UDI-non-UMI Adapters 10bp - 3 digit (eg: IDT_10nt_UDI_i7_001-IDT_10nt_UDI_i5_001)
- Integrated DNA Technologies (IDT) - xGen UDI-UMI Methylated Adapters 8bp - 2 digit (eg: IDTMU701-IDTMU501)
- Lucigen - NxSeq AmpFREE Low DNA Library Kit (eg: NS_Adaptor_1)
- MGI Tech - MGI Easy DNA Adaptors (eg: MGI01_A01_Barcode_1)
- MGI Tech - MGI UDB dual index barcodes (eg: MGI_Barcode1_UDB_001-MGI_Barcode2_UDB_001)
- MGI Tech - Single index balanced pools (eg: SI-MGIPool-C01A.4plex)
- New England Biolabs (NEB) - NEBNext Dual index Multiplex Oligos for Illumina Index Primers Set 2 (NEB #E7780S) (eg: NEBNextDi701-NEBNextDi501|NEBNextDi502|...|NEBNextDi516)
- New England Biolabs (NEB) - NEBNext Enzymatic Methyl-seq (eg: NEBEM701-NEBEM501)
- Oxford Nanopore - Native Barcoding Set (eg: BC01)
- Perkin Elmer - Bioo Scientific NEXTflex-96 DNA Barcodes (eg: BioOHT_1)
- QIAgen - QIASeq Direct UDI Plates A/B/C/D/E/F/G/H (eg: QIAUDI7001-QIAUDI5001)
- QIAgen - QIAseq miRNA NGS 96 Index IL (QIAgen #331565) (eg: QIASeq_miRNA_MIHT01)
- QIAgen - QIAseq miRNA NGS ILM IPD (eg: QIAmiRNA_IDP01)
- Universal Sequencing - TELL-Seq index (eg: T501)
- Custom index - Institut de recherches cliniques de Montreal (IRCM) (eg: D701rc-D501|D502|...|D508)
- Custom index - Kaestner Lab ATAC-seq Index (version 2) (eg: v2_Ad2.1-v2_Ad1.1)
- Custom index - McGill DAI Genome-wide CRISPR screening (eg: DAIcrisper_R01)
- Custom index - Made by IDT for McGill - xGen UDI-UMI Like Methylated Adapters (eg: Methyl_TS_UDI_UMI_IDT8nt_i7_01-Methyl_TS_UDI_UMI_IDT8nt_i5_01)
### Sample Manifest - New update script related changes:
- In hidden range tab: Column changes to allow pasting of output from script.
- Updated excel INDIRECT lookups to accept ranges defined by script.
- LIMS index definitions and library type now contains a "publish" column for each kit name and library type allowing the use of items in the system without adding them to the manifest.
### LIMS Configuration
- __sample reception__, __all index validation__ - Longer list of library type including description and library category (from enseqlopedia.com). All made compatible with Illumina run processing and LIMS.  Full list of library type is defined in lims but not published in full in this version of the manifest.
- __sample reception__, __all index validation__ - Adding index kit name and publish field to adapter type file.
- __sample reception__ - Added check to force Species field to be defined.
- __sample reception__ - Fixed bug to ensure that both index in dual index are verified.
- __sample reception__, __all index validation__ - Adding SHARE-Seq (ATAC, RNA) library type to definitions.
- __sample reception__, __all index validation__ - Adding new custom (by IDT for McGill) Methylated adapter set "with shifted pairing".
- __all index validation__ - Adjustment to allow non unique i7 entries in adapter types.
### Illumina Run Processing
- Adjustment to allow non unique i7 entries in adapter types.
- Adjustment to ruby script to handle added columns to defintions) for sequenced index validation definitions.
### MGI Run Processing
- Version in production (genpipes/ddb7b96) is compatible with this version of adapter index definitions.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/3.0.0%0D2.1.2#diff)

## [2.1.2] - 2022-06-22
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding library type SHARE-Seq_ATAC and SHARE-Seq_RNA.
### Run Processing Index Definitions
- Same as for lims configuration.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/2.1.2%0D2.1.1#diff)

## [2.1.1] - 2022-06-21
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding Dovetail HiChIP MNase Kit index (8 index, 8-8nt Dual index).
### Run Processing Index Definitions
- Adding Dovetail HiChIP MNase Kit index (8 index, 8-8nt Dual index).
###
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/2.1.1%0D2.1.0#diff)

## [2.1.0] - 2022-06-17
### Run monitor
- Changing source for date field so that MGIG400 scratch processing directory matches the final processing directory.
### Run Processing (Illumina)
- Adjusting top blast hit detection rule to exclude non-human species still with "human" in the name.
### Run Processing (Nanopore-covid)
- Update to covseq tools to address some ncov_tools jobs crashes.
- Updating covid-19 ONT (Oxford Nanopore Technologies) run processing pipeline (genpipes branch:master/4.2.0).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/2.1.0%0D2.0.1#diff)

## [2.0.1] - 2022-06-08
### LIMS Configuration
- __sample reception__, __all index validation__ - Completing IDTMU (IDT methylated dual index 8-8 for WGBS) for range 1-48.
### Run Processing
- Completing IDTMU (IDT methylated dual index 8-8 for WGBS) for range 1-48.
###
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/2.0.1%0D2.0.0#diff)

## [2.0.0] - 2022-06-02
### Run Monitoring
- Adding MGIG400 and ONTCOVID run processing to mini-dashboard links.
### Run Processing (run monitor)
- Adding CentrEau (FRI1621) and CoVRNet_LF (FLA1601) to Illumina rapid response emails.
### Run Processing (MGI/G400)
- Added variables in run processing end email table (nb_bases, duplicate_rate and avg_qual).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/2.0.0%0D1.7.1#diff)

## [1.7.1] - 2022-05-25
### LIMS Configuration
- __genotyping__ - Modified robocopy file generation script to retrieve project names and IDs when the API returns more than one page.
### Run Processing
- Adding Mina to CovSeQ/nanopore run processing emails.
- Removing Mathieu from rapid response emails.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.7.1%0D1.7.0#diff)

## [1.7.0] - 2022-05-16
### LIMS Configuration
- __sequencing__ - Added "Barbara McClintock (A01861)" in the drop down in the "Cluster Generation (NovaSeq) McGill" step. (added manually in prod 2022-05-13).
### Run Processing
- Adding third genome center NovaSeq to the run processing configuration (instrument_list.csv) (Instrument ID: A01861 aka Barbara McClintock). 
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.7.0%0D1.6.1#diff)

## [1.6.1] - 2022-05-12
### LIMS Configuration
- __sample reception__,__all index validation__ - Correction to NEBNext Single index Oligos Set 2 and 4 (all these index should be 6bp).
### Run Monitor
- Adjusted report table in final email to display new variable names in genpipes json.
### Run Processing (MGI/G400)
- Changes to genpipes in new commit genpipes/ddb7b96 from previous commit genpipes/12b85e8 covers issues with dual index libraries sequenced as dual index or as single index.
- Fixed job throttling so that each job has its own scheduler log file.
- Fixed report jobs
- Fixed verifyBamID job dependencies + updated to verifyBamID2
- Adjust fastq header conversion for fgbio 1.4
- Rebased with GenPipes “dev” branch
- See all the genpipes changes here: https://bitbucket.org/mugqic/genpipes/branches/compare/ddb7b96%0D12b85e8#commits
### Run Processing (Illumina)
- Correction to NEBNext Single index Oligos Set 2 and 4 (all these index should be 6bp).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.6.1%0D1.6.0#diff)

## [1.6.0] - 2022-04-08
### LIMS Configuration
- __sample reception__,__all index validation__ - Modifying QIASeq_miRNA_MIHT indexes to use the mirna_qiagen backbone.
- __library step__ - Added striptube container option for CaliperGX step inside 10x single cell workflow. (added manually in prod 2022-04-06).
- __library step__ - Added 96 well plate container option for Library Pool inside MGI Conversion Protocol. (added manually in prod 2022-04-08).
### Run Processing (Illumina)
- Added temporary file cleanup steps to Illumina Run processing pipeline to use a smaller data footprint for scratch.
- Added TELL-Seq library support where the index is in i5 and the molecule barcode is in i7.
- Modified index validation ruby script to include MGI in the index lists produced (barcodes_by_sequence.*).
- Included the generation of the lists above to the deployment procedure. The ruby command will automatically be included in DEPLOYMENT.md whenever index definitions are changed.
### Run Processing (Nanopore)
- Added qos:high option controlled by project id for nanopore-covid processing to improve processing response time (run monitor will only use this for project CoVSeQ initially).
### Support Utilities
- Modified support_utilities/sendmailscript.sh that generates DEPLOYMENT.md to include above ruby call in deployment script when needed to generate barcodes_by_sequence.* at deploy time.
- Added script (runprocessing/bclcount.sh) to help validate if the raw run was copied properly from the sequencer before run processing is started. To be used when the Illumina Universal Copy Service (UCS) fails.
- Modified runprocessing/restartevent.sh to include a "fast" option to skip job id listing. Included delete commands for /lb/robot and /nb/Research to facilitate clean restarts.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.6.0%0D1.5.4#diff)

## [1.5.4] - 2022-03-17
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding "GenDx NGSgo IndX plate I,II,III and IV (cat. no. 2342456)" to index definitions.
### Run Processing
- Adding "GenDx NGSgo IndX plate I,II,III and IV (cat. no. 2342456)" to index definitions.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.5.4%0D1.5.3#diff)

## [1.5.3] - 2022-03-11
### LIMS Configuration
- __sequencing__ - Modified event file script to retrieve project names and IDs when the API returns more than one page.
- __sample reception__, __all index validation__ - Adding "QIAseq miRNA NGS 96 Index IL (cat. no. 331565)" to index definitions.
- __library step__ - Added "QIAseq miRNA NGS 96 Index IL (cat. no. 331565)" to RNASeq library step. (added manually in prod 2022-03-09).
### Run Monitor
- Adding genpipe processing start success/failure detection and email for MGI-G400.
- Adding job status table and number of job complete (Vs total job) in final MGI-G400 email.
- Adapted pipeline call to new version of genpipes MGI run processing script.
- Added T7 run detection to monitor.
- Changes to genpipes in new commit 12b85e8 from previous commit 870b43c covers improvements and issues such as dual index demultiplexing, mouse skipping verifybamid step, multiQC compatibility and command option mgig400/mgit7 processing.
- See the genpipes changes here: https://bitbucket.org/mugqic/genpipes/branches/compare/12b85e8%0D870b43c30364
- Removing Pierre Lepage for run completion emails (Illumina/MGI)
- Set the cores:threads to 8:4 to reduce overall run time and ensure that ONT-covid report is more reliable. (done manually in prod 2022-03-11)
### Run Processing
- Adding "QIAseq miRNA NGS 96 Index IL (cat. no. 331565)" to index definitions.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.5.3%0D1.5.2#diff)

## [1.5.2] - 2022-02-23
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding TELL-Seq library type and TELL-Seq index T501-T524.
- __library step__ - Added IDT for illumina-Nextera UDP V1 added to RNASeq library step. (added manually in prod 2022-01-26).
- __library step__ - Added QIAgen index (plates A-F) to VirusSeq steps (multiple workflows). (added manually in prod 2022-02-18).
- __library step__ - Added QIAgen index definition (plates E-F-G-H) to adapter definitions.
### Run Monitor
- Adding Tony Tyr for run completion emails (Illumina/MGI)
### ONT-covid Run Processing
- Reduced report generation cores in genpipes ini to reduce overall run time. (done manually in prod 2022-02-16)
- Adding ARTIC v4.1 primer case to PCR Kit Name genpipe ini lookup in the case where the Kit name is not defined. (added manually in prod 2022-01-29).
- Adding restartevent-ONT.sh script to the code repository (script facilitates ONT-covid run restarts).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.5.2%0D1.5.1#diff)

## [1.5.1] - 2022-01-21
### LIMS Configuration
- __sequencing__ - Adding unset variable to event file script affecting external library.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.5.1%0D1.5.0#diff)

## [1.5.0] - 2022-01-14
### LIMS Configuration
- __sequencing__ - Modified event file script to output sequencer type, lab run counter and PCR primer names (Artic).
### Run Monitor
- Modified run monitor to split ONT-Covid processing by project and by PCR primer version.
- Created pcr_kit_name_lookup.txt file to easily map new Artic primer versions to corresponding ONT-Covid processing ini file.
### Run Processing
- Added illumina run processing cleanup utility script to runprocessing folder.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.5.0%0D1.4.2#diff)

## [1.4.2] - 2021-12-07
### LIMS Configuration
- __all index validation__ - Correction to cut and tag version 2 index definitions (Epicypher protocol v1.5).
- __all index validation__ - Adding NEBNext Multiplex Oligos for Illumina (Index Primers Set 1) E7335.
### Run Monitor
- Added Covarrnet_Calgary (PAR1425) to RAPIDRESPONSE list (added manually in prod 2021-12-07)
### Run Processing
- Correction to cut and tag version 2 index definitions (Epicypher protocol v1.5).
- Adding NEBNext Multiplex Oligos for Illumina (Index Primers Set 1) E7335.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.4.2%0D1.4.1#diff)

## [1.4.1] - 2021-11-22
### LIMS Configuration
- __all index validation__ - Adding cut and tag version 2 index definitions (Epicypher protocol v1.5).
- __library step__ - Added workflow for MGI PCR-free with T7 named 360_MGIPCRfree_T7_1 (added manually in prod 2021-11-22).
### Run Processing
- (Run Monitor) Removed Reply-to hercules for ONT final email to facilitate team conversation and not create tickets.
- (Run Monitor) Correcting issue with initial advanced notice email for rapid response / illumina runs.
- (Run Monitor) Modified run monitor readset creation to better handle the case where summary files are not output from sequencer (using basecalling_enabled field from ONT final_summary).
- (Run Monitor) Modified run monitor email text when detecting illumina run issues.
- Adding cut and tag version 2 index definitions (Epicypher protocol v1.5).
- Updating index validation files according to new indexes.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.4.1%0D1.4.0#diff)

## [1.4.0] - 2021-11-09
### LIMS Configuration
- __sequencing__ - Created an MGI sequencing protocol for the DNBSeq-T7.
- __library step__ - Added Artic primer version to dropdown in VirusSeq PCR (added manually in prod 2021-11-11).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.4.0%0D1.3.0-rc7#diff)

## [1.3.0] - 2021-11-04
### LIMS Configuration
- __all index validation__ - Modification of adapter settings files to include information about technology (Illum./MGI/ONT)
- __all index validation__ - Modification of adapter settings files to include information about i7/i5 index pairing for dual index (unique pair versus combinatorial).
### Run Processing
- Addition of ruby script to convert new adapter settings files for "CountIlluminaBarcodes".
- Improvement to illumina run processing index report to list the unexpected index by name (new index report is embedded in the main illumina run report).
- Modification to run monitor to ensure that all the jobs have completed, (prepare_report not always last) for covid-19 ONT pipeline (branch:covseq_ont/2283d9c).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.3.0-rc7%0D1.2.0#diff)

## [1.2.0] - 2021-09-14
### LIMS Configuration
- __all index validation__ - Adding Broad sgRNA (guide RNA) index definition.
### Run Processing
- Adding Broad sgRNA (guide RNA) index definition.
- Adding integration to automatically start the covid-19 ONT (Oxford Nanopore Technologies) run processing pipeline (branch:covseq_ont/2283d9c).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.2.0%0D1.1.1#diff)

## [1.1.1] - 2021-09-14
### LIMS Configuration
- __library step__ - Added workflow for scATAC for the MGI DNBSeq-G400.  (added manually in prod 2021-09-02).
- __all index validation__ - Correction to FeatureBarcode adapter backbone definition.
- __all index validation__ - Correction to Swift barcode sequence definition.
### Run Processing
- Correction to FeatureBarcode adapter backbone definition.
- Correction to Swift barcode sequence definition.
### Sample Manifest
- Adding logic in manifest to detect invalid reference entered, eg. Homo_sapiens:GRCh39,40,41.
- Adding logic in manifest to detect missing number in pool value for external pools submitted.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.1.1%0D1.1.0#diff)

## [1.1.0] - 2021-08-31
### LIMS Configuration
- __all index validation__ - Adding Ovation SoLo RNA-Seq index definition.
- __all index validation__ - Adding Perkin Elmer BioO NextFlexChIPSeq index definition.
### Run Processing
- Adding Ovation SoLo RNA-Seq index definition.
- Adding Perkin Elmer BioO NextFlexChIPSeq index definition.
### Sample Manifest
- Adding Ovation SoLo RNA-Seq index to manifest.
- Adding Perkin Elmer BioO NextFlexChIPSeq index to manifest.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.1.0%0D1.0.1#diff)

## [1.0.1] - 2021-08-18
### Run Processing
- Correcting issue with generating Read 2 when I2 is sequenced as a read instead of used for dual index demultiplexing.
### Sample Manifest
- Correcting issue with Swift index dropdown.
- Adding "Non-listed commercial indexes" in the index dropdown.
- Moving "Custom Index" to the bottom of the dropdown.
- Adding text: Prior to submitting your workorder, please contact pm.genome@mcgill.ca to discuss your custom or non-listed indexes.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.0.1%0D1.0.0#diff)

## [1.0.0] - 2021-07-28
### LIMS Configuration
- __library step__ - Adding protocol for Nanopore library QC.
- __library step__ - Adding protocol for Nanopore Sequencing (protocol supports MinION, GridION, Flongle and PromethION).
- __library step__ - Added step variable to adjust maximum sample volume for the VirusSeq Normalization step. (added manually in prod 2021-07-30).  
- __library step__ - Mopdified script to use a step variable to adjust maximum sample volume for the VirusSeq Normalization script to accomodate Illumina and Nanopore.
### Run Processing
- New nanopore protocols remain compatible with script generating event file for run processing.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/1.0.0%0D0.6.20#diff)

## [0.6.20] - 2021-07-14
### LIMS Configuration
- __all index validation__ - Bug fix to index pooling checking logic affecting 10x genomics single index.
### Run Processing
- Bug fix to run processing demultiplexing logic affecting 10x genomics single index.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.20%0D0.6.19#diff)

## [0.6.19] - 2021-07-07
### LIMS Configuration
- __all index validation__ - Improvement to index pooling checking logic to handle nested index names.
### Run Processing
- Adding projects to run processing rapid response list. EQAP (RAG1282) and Covid_TechDev (RAG1283) added with same recipients as for CoVSeQ, Epocrates and Covid_Wastewater.
- Improvement to run processing demultiplexing logic to handle nested index names.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.19%0D0.6.18#diff)

## [0.6.18] - 2021-06-29
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding adapter definition for 10x Genomics Dual Index (NN series).
- __library step__ - Adding 10x Genomics Dual Index (NN series) to scRNASeq step (added manually in prod 2021-06-29).
- __library step__ - Adding MGI UDB dual index barcodes (1-196) and MGI UDB dual index barcodes (97-192) to PCR-Free MGI, RNASeq MGI and WGBS MGI  (added manually in prod 2021-06-29).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.18%0D0.6.17#diff)

## [0.6.17] - 2021-06-22
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding adapter definition for NEBNext Unique dual index (non UMI) Set1,2,3,4 sortable by columns for LIMS.
- __sample reception__, __all index validation__ - Adding adapter definition for MGISeq Unique dual index (ready for external library submission).
- __library step__ - Adding NEBNext Unique dual index (non UMI) Set1,2,3,4 to RNASeq step (added manually in prod 2021-06-22).
- __library step__ - Adding NEBNext Unique dual index (non UMI) Set1,2,3,4 to PCR-enriched step (added manually in prod 2021-06-22).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.17%0D0.6.16#diff)

## [0.6.16] - 2021-06-14
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding adapter definition for IDT TruSeq Unique Dual Index with 10-10 cycles.
- __library step__ - Added IDT TruSeq Unique Dual Index Plate 1,2,3 and 4 to PCR-free step (added manually in prod 2021-06-14).
- __library step__ - Added IDT TruSeq Unique Dual Index Plate 1,2,3 and 4 to PCR-encriched step (added manually in prod 2021-06-14).
- __library step__ - Added IDT TruSeq Unique Dual Index Plate 1,2,3 and 4 to RNASeq step (added manually in prod 2021-06-14).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.16%0D0.6.15#diff)

## [0.6.15] - 2021-06-04
### LIMS Configuration
- __sample reception__, __all index validation__ - adding 10x Genomics Single Cell Visium Spatial RNA library type. (correcting typo)
- __sample reception__, __all index validation__ - Adding NEBNext Multiplex Oligos for Illumina Index Primers Set 1,2,3,4
### Run Processing
- See above library type definition
- Updated run processing restart script: Speed improvement for listjobs.sh utility.
- Updated run processing restart script: More informative restart_event.sh utility now showing the run monitor process id.
### Sample Manifest
- Adding a tab for examples of how to fill the manifest.
- Adding a tab to help define new custom indexes.
- Hiding tab containing dropdown information.
- Adding index series to manifest: GeoMx DSP index 1-768 (renamed "GeoMx DSP index (subset)")
- Adding index series to manifest: NEBNext Unique dual index (non UMI) Set1 (NEB #E6440)  (renamed "NEBNext Unique dual index (non UMI)")
- Adding index series to manifest: NEBNext Unique dual index (non UMI) Set2 (NEB #E6442)
- Adding index series to manifest: NEBNext Unique dual index (non UMI) Set3 (NEB #E6444)
- Adding index series to manifest: NEBNext Unique dual index (non UMI) Set4 (NEB #E6446)
- Adding index series to manifest: NEBNext Multiplex Oligos for Illumina Index Primers Set 1,2,3,4 (NEB #E7335, #E7500, #E7710, #E7730)
- Adding index series to manifest: 10x Genomics Dual index NT series
- Adding index series to manifest: IDT Nextera Compatible UDI
- Adding index series to manifest: QIAgen miRNA
- Adding index series to manifest: Switf adapter
- Adding index series to manifest: IBIS custom combinatorial dual index
- Adding library types to manifest: 10x Genomics Single Cell Visium Spatial RNA
- Adding library types to manifest: 10x Genomics Single Cell Feature Barcode
- Adding library types to manifest: 10x Genomics Single Cell ATAC
- Adding library types to manifest: 10x Genomics Single Cell V-D-J
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.15%0D0.6.14#diff)

## [0.6.14] - 2021-06-03
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding GEOmx index definitions (range 97-768 unique dual index).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.14%0D0.6.13#diff)

## [0.6.13] - 2021-05-26
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding GEOmx index definitions (96 unique dual index).
- __sample reception__, __all index validation__ - Adding NEB Next index definitions for set 2,3 and 4 (96 unique dual index each).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.13%0D0.6.12#diff)

## [0.6.12] - 2021-05-12
### Run Processing
- Integrated new mgi run processing with run/event monitor (https://bitbucket.org/mugqic/genpipes/commits/870b43c).
- MGI pipeline made compatible with event file.
- MGI pipeline outputs modified fastq header format (compatible with 10x Genomics cellranger v5+).
- MGI pipeline automatically copies the event file into the run processing folder to facilitate mgi report generation and copy to genap.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.12%0D0.6.11#diff)

## [0.6.11] - 2021-05-10
### LIMS Configuration
- __index pool validation__ - Updated the index pooling check code to handle the TruSeq/Nextera backbone used by 10x Feature Barcode libraries in the case where dual index is used to demultiplex the lane.
- __sample reception__, __all index validation__ - Adding IDT index definition (sold by IDT) that are Nextera compatible.
### Run Processing
- Modified run processing pipeline to output index 2 unused by demultiplexing as a read (named I2) regardless of the library type. This is to support HaloPlex and 10x multiome kits (scATAC).
- Modified run processing pipeline to handle TruSeq/Nextera library structure used by 10x Genomics Feature Barcodes.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.11%0D0.6.10#diff)

## [0.6.10] - 2021-04-23
### LIMS Configuration
- __sequencing__ - Added a novaseq sequencer name dropdown in the novaseq clustering step. (added manually in prod 2021-04-23).
### Run Processing
- Changed the first novaseq name to Rosalind Franklin from NovaScotia.
- Added second genome center novaseq to the run processing configuration (Instrument ID: A01433 aka Carrie Derick).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.10%0D0.6.9#diff)

## [0.6.9] - 2021-04-20
### LIMS Configuration
- __sample reception__, __all index validation__ - adding 10x Genomics dual NT set indexes.
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.9%0D0.6.8#diff)

## [0.6.8] - 2021-04-13
### LIMS Configuration
- __sample reception__, __all index validation__ - adding custom dual index sequences from IBIS.
- __library step__ - Added workflow for PCR-free combined with MiSeq (added manually in prod 2021-04-13).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.8%0D0.6.7#diff)

## [0.6.7] - 2021-04-01
### LIMS Configuration
- __library step__ - Correction to QIAgen miRNA adapter definition.
- __sample reception__, __all index validation__ - adding QIAgen unique dual index definitions.
### Run Processing
- See above adapter definitions
- Adding project FRI1138 (Covid_Wastewater) to RAPID RESPONSE list with the email of users involved in that project.
- Adding project QUA1139 (Epocrates) to RAPID RESPONSE list with the email of users involved in that project.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.7%0D0.6.6#diff)

## [0.6.6] - 2021-03-23
### LIMS Configuration
- __library step__ - Added QIAgen miRNA adapter definition to VirusSeq.
- __library step__ - Added QIAgen miRNA index to RNASeq step (added manually in prod 2021-03-23).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.6%0D0.6.5#diff)

## [0.6.5] - 2021-03-15
### LIMS Configuration
- __library step__ - Added Switf adapter definition to VirusSeq.
- __library step__ - Added Swift index to VirusSeq step (added manually in prod 2021-03-19).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.5%0D0.6.4#diff)

## [0.6.4] - 2021-03-04
### LIMS Configuration
- __sample reception__, __all index validation__ - adding NEBNext UMI (unique dual index) library index Kit definition.
- __sample reception__ - Small modification to sample reception script to handle new manifest and remain backward compatible with old manifests.
### Run Processing
- See above adapter definitions
### Sample Manifest
- Allowing text typed in grayed out fields to be visible in red strikethrough.
- Adding index series to manifest: Twinstrand, NEBNext UDI, Geo MX DSP and Agilent Sureselect XT (384)
- Deleting hidden row 1
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.4%0D0.6.3#diff)

## [0.6.3] - 2021-01-25
### Run Processing
- Integration of genpipes MGI run processing pipeline.  Run monitor was modified to handle illumina runs and MGI runs. Run monitor will send emails at the end of run processing for MGI also.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.3%0D0.6.2#diff)

## [0.6.2] - 2021-01-21
### LIMS Configuration
- __library step__ - Changing the maximum volume for virus seq normalization to 7.5uL.  Fixing issue allowing variable total volume.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.2%0D0.6.1#diff)

## [0.6.1] - 2021-01-06
### LIMS Configuration
- __library step__ - Adding sureselect HT V2 range from 97-384 to PCR-enriched library prep step, created 4 sets (ABCD) for each group of 96 (added manually in prod 2021-01-05).
- __library step__ - Adding Noindex option to VirusSeq step to accomodate nanopore libraries done with no index.  Adding extra protocol.io and Kit Name options to VirusSeq step (added manually in prod 2021-01-05).
- __adapters__ - Adding IDT adapters and renamed nanopore adapter definitions.
### Run Processing
- Modification to illumina run processing to use 8 or 8-8 cycles to demultiplex if 10 or 10-10 cycles was used on the flowcell and all the libraries in the lane are 8 or 8-8 cycles. If mixed index length libraries are pooled, demultiplexing is performed on longest cycles.
### Sample Manifest
- Updated manifest with automatic highlighting of required columns based on sample type to facilitate use.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.1%0D0.6.0#diff)

## [0.6.0] - 2020-12-15
### LIMS Configuration
- __management step__ - Created new step allowing batch file driven data curation for metadata on steps, artifacts, index and container naming.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.6.0%0D0.5.1#diff)

## [0.5.1] - 2020-12-01
### LIMS Configuration
- __library step__ - Added Sureselect index to PCR-enriched (added manually in prod 2020-11-27).
- __library step__ - Added surelect options for Capture step drop down (added manually in prod 2020-11-27).
### Sample Manifest
- Small correction to sample manifest, removing whitespace in carrier columns.
### Run Processing
- Changed wording in alert emails from "Network issue" to "Sequencer writing issue" since the majority of cases are caused by a sequencer error.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.5.1%0D0.5.0#diff)

## [0.5.0] - 2020-11-20
### LIMS Configuration
- __library step__ - Created: MGI PCR-free protocol.
- __library step__ - Created an MGI RNASeq protocol.
- __library step__ - Created an MGI WGBS protocol.
- __library step__ - Created an MGI library Conversion protocol.
- __library qc__ - Created an MGI library QC protocol.
- __sequencing__ - Created an MGI sequencing protocol for the DNBSeq-G400.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.5.0%0D0.4.15#diff)

## [0.4.15] - 2020-11-19
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding custom adapters for client D705rc to D712rc (D701rc to D704rc already exist).
### Run Processing
- Adding custom adapters for client D705rc to D712rc (D701rc to D704rc already exist).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.15%0D0.4.14#diff)

## [0.4.14] - 2020-11-16
### LIMS Configuration
- Added extra step in VirusSeq protocol to handle cherry picking of samples with program for robot.
- Added concentration to queue view and icebucket for VirusSeq Normalization step.
- Added actual protocol.io links in dropdowns.
- Added Nanopore support for library part inside VirusSeq itself.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.14%0D0.4.13#diff)

## [0.4.13] - 2020-11-13
### Run Processing
- Fixed issue with readset file generated for rapid reponse.
- Added LibraryType to readset file generated for rapid reponse.
- Adding Francois Levebvre and Rob Syme to all rapid response emails for monitoring.
- Changed the run processing complete reply-to email to hercules@mcgill.ca.
- Modified report section of run processing to remove commas and dots from blast result names.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.13%0D0.4.12#diff)

## [0.4.12] - 2020-11-05
### LIMS Configuration
- __index pool validation__ - Fixed an issue with index generation.  Problem was introduced when modifying the code to support 10x dual index and affects 10x single index.
### Run Processing
- Fixed an issue with index generation.  Problem was introduced when modifying the code to support 10x dual index and affects 10x single index.
- Removing Alfredo Staffa from email lists.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.12%0D0.4.11#diff)

## [0.4.11] - 2020-10-21
### LIMS Configuration
- __sequencing__ - Fixed an issue of sample ordering in the script that generates a csv during Library Normalization for NovaSeq and HiSeqX
### Run Processing
- Added force to symlink creation for index step to avoid having jobs failures.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.11%0D0.4.10#diff)

## [0.4.10] - 2020-10-16
### Run Processing
- Adding Elizabeth Caron to run processing complete emails (to facilitate run validation)
- Fixed issue with verifybamid running on non-human datasets and systematically failing.
### Sample Manifest
- Small correction to sample manifest, removing whitespace in Sex column.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.10%0D0.4.9#diff)

## [0.4.9] - 2020-10-09
### LIMS Configuration
- __sample reception__, __all index validation__ - Correction (no. 1) for DSP dual index adapters (subset).
### Run Processing
- Modified run processing config files to handle 10x dual index and the pool of four i7 index in combination with NovaSeq.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.9%0D0.4.8#diff)

## [0.4.8] - 2020-10-07
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding TwinStrand dual index adapters.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.8%0D0.4.7#diff)

## [0.4.7] - 2020-09-29
### Run Processing
- Changed the blast version to 2.10.  With this version blast are always points to the latest version of nr available on cvmfs. Includes "severe acute respiratory syndrome coronavirus 2".
- Modified run processing code to handle 10x dual index as opposed to a pool of four i7 index.
- Corrected issue in event monitor script to wait for some metadata to be written to disk before creating the readset files for rapid reponse.
- Adjusted the blast species matching to only use the first two keywords if the submitted species name is too long.
### LIMS Configuration
- __all index pooling QC__ - Modified library pooling code to handle 10x dual index as opposed to a pool of four i7 index
- __sample reception__, __all index validation__ - Adding DSP dual index adapters (a subset).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.7%0D0.4.6#diff)

## [0.4.6] - 2020-09-23
### Run Processing
- Added ability to handle demultiplexing from Novaseq runs with version 1.5 reagents while remaining backwards compatible with old reagents.
### Sample Manifest
- Added "Viruses:Severe acute respiratory syndrome coronavirus 2 (Taxon ID:2697049)" to species dropdown.
- Added the following index series to the dropdowns: 10x Genomics Dual Index
- Added the following index series to the dropdowns: Illumina IDT Nextera UDP V1
- Added the following index series to the dropdowns: Illumina IDT Nextera UDP V2
- Added the following index series to the dropdowns: Illumina IDT TruSeq UDI V1
- Added the following index series to the dropdowns: Illumina IDT TruSeq UDI V2
- Added the following index series to the dropdowns: Sure Select HT Low Input
- Added the following index series to the dropdowns: SeqWell Indexes
- Added the following index series to the dropdowns: Schloss Indexes
- Extended the following index series in the dropdowns: Agilent SureSelect XT V2
- Minor edits to instructions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.6%0D0.4.5#diff)

## [0.4.5] - 2020-09-17
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding Schloss 16S rRNA adapters index for external library.
### Run Processing
- See above for adapter settings and adapter types change.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.5%0D0.4.4#diff)

## [0.4.4] - 2020-09-05
### LIMS Application
- Enabling the option for users (Facility Administrator only) to edit lims metadata after step is completed for data curation.
### Run Processing
- Adding project RAG936 (CoVSeQ) to RAPID RESPONSE list with the email of users involved in that project.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.4%0D0.4.3#diff)

## [0.4.3] - 2020-08-24
### LIMS Configuration
- __library step__ - Adding UDP V1 Illumina-Nextera Unique Dual Index adapters to PCR-free steps.  UDP IDT Illumina Nextera V1 and V2 already in adapters settings.
- __library step__ - Adding UDI V1 to PCR-free, removing UDI V2. UDI IDT Illumina TruSeq V1 and V2 already in adapters settings.
- __axiom genotyping__ - Corrected issue with 4 digit project code (PRA1005) affecting "Scanning QC step" step.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.3%0D0.4.2#diff)

## [0.4.2] - 2020-08-12
### LIMS Configuration
- __axiom genotyping__ - Corrected issue with 4 digit project code (PRA1005) affecting "Denaturation and Hybridization McGill 1.0" step.
### Run Processing
- Corrected bug where alignment metadata was being assigned to samples with no reference provided.
### LIMS metadata reports
- Adding MiSeq and iSeq into the "by project" metadata reports used by the lab and the managers.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.2%0D0.4.1#diff)

## [0.4.1] - 2020-08-10
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding custom index for client (DAIcrisper_R01 - DAIcrisper_R10).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.1%0D0.4.0#diff)

## [0.4.0] - 2020-07-27
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding Illumina UDP V1 adapter definitions, naming for Clarity.
- __library step__ - Created VirusSeq Protocol to use with Coronavirus Sequencing in Quebec, (CoVSeQ) project.
- __library step__ - Added VirusSeq to library definitions.
- __sequencing__ - Modified event file script to handle case external lib -> capture-> pooling-> replicas.
- __sequencing__ - Modified event file script to speed it up when encoutering large in lims pools or captures.
### LIMS Application
- Adding the option for users (Reseacher) to move/remove samples from the step queues.
- Removed the option for users (Reseacher) to create/modify samples without using the official sample reception step.
- The above operation removes the manifest download option, therefore the manifest will be emailed along side release notes at every update.
### Repository
- Moved unused protocol definitions to new folder other_slicer_snapshots.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.4.0%0D0.3.11#diff)

## [0.3.11] - 2020-07-06
### Run Processing
- Adding RAPID RESPONSE functionality to event monitor.
- Adding bravolims-qc bravolims-dev awareness to run processing and event monitor for emulation in QC and DEV environments.
- Adjusting QOS flag to disable for dev and qc.
- Adding restartevent.sh and listjobs.sh run processing management utilities.
- Adding config file to manage email lists for run processing alerts, completion and rapid response.
- Fix run processing (Pct Chimera/Pct Adapter) were not parsed correctly.
- Fix I2 fastq was not being copied for RNA type protocols.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.11%0D0.3.10#diff)

## [0.3.10] - 2020-07-01
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding SureSelect XTHSv2 adapters (1-16 exist, adding 17-96).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.10%0D0.3.9#diff)

## [0.3.9] - 2020-06-16
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding UDPV1 (Nextera) and UDIV1 (TruSeq) V1 adapters.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.9%0D0.3.8#diff)

## [0.3.8] - 2020-06-11
### LIMS Configuration
- __sample reception__, __all index validation__ - Correcting UPD V2 index series and UDI V2 index series.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.8%0D0.3.7#diff)

## [0.3.7] - 2020-06-10
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding SeqWell index series to adapter definitions.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.7%0D0.3.6#diff)

## [0.3.6] - 2020-06-06
### LIMS Configuration
- __library step__ - Adding UDI Illumina-TruSeq Unique Dual Index adapters to PCR-free and PCR-enriched steps.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.6%0D0.3.5#diff)

## [0.3.5] - 2020-06-05
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding UPD V2 index series and UDI V2 index series to adapter definitions.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.5%0D0.3.4#diff)

## [0.3.4] - 2020-06-04
### LIMS Configuration
- __sample reception__, __all index validation__ - Correction to Illumina UDP index series.
- __sample reception__ - Bug fix for sample reception of pool.
- __sequencing__ - Modified event file script to output correct lane fraction for external libraries.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.4%0D0.3.3#diff)

## [0.3.3] - 2020-05-21
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding Illumina UDP index series. Removing duplicate index definitions.
- __sequencing__ - Modified event file script to properly output event file in case a replicate of capture is created in the LIMS. 
### Run Processing
- Modifying run processing to handle 10x Genomics pooled index based on the index name instead of library type name.
- Ensuring that 5K reads are used for blast at least or all reads if less are available.
- Removing carriage return from event file provided to handle cases where it is introduced manually.
- If sex is not specified replacing empty value by 'N/A' to avoid flagging a sex match error. 
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.3%0D0.3.2#diff)

## [0.3.2] - 2020-05-04
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding Amplicon DNA/RNA to library_protocol_list.csv for sample reception.
### Run Processing
- See above for library_protocol_list.csv.
### Sample Manifest
- Adding Amplicon DNA/RNA to the manifest.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.2%0D0.3.1#diff)

## [0.3.1] - 2020-04-15
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding adapter structure definition for 10X Genomics Visium.
- __sample reception__, __all index validation__ - Adding adapter structure definition for 10X Feature Barcode.
- __sample reception__, __all index validation__ - Adding adapter structure definition for Illumina RPI (small RNA).
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.1%0D0.3.0#diff)

## [0.3.0] - 2020-03-10
### LIMS Configuration
- __sample QC__ - Added a Tapestation RNA Qc Protocol.
- __sample QC__ - Adding steps for manual image attachment for Femto Pulse and Tapestation for DNA.
- __sample QC__ - Show Carrier Coordinate in the DNA QC Tecan and Gel.
- __sequencing__ - Added support for the submitted library replicas
- __sequencing__ - Added info from Capture step for submitted libraries to ru processing event file.
- __sequencing__ - Added info Target Cells UDFs added to the event file. 
- __library step__ - Split ChIP-seq and ChIPmentation steps into 2 protocols for full worklflow version.
- __library step__ - Adding robot normalization step to WGBS (same as PCR-free and enriched).
- __library qc__ - Enabled autoplacements, 96->96 and 96->384, group by project and show workflow in library batch step.
- __sequencing__ - Moving Library Normalization csv driver file and library volume update to Library Normalization step for HiSeq X and NovaSeq protocols.
- __library step__ - Added "10x Genomics Single Cell ATAC".
- __library step__ - Added "10x Genomics Single Cell RNA QC" to use with full workflows.
- __library step__ - Added "10x Genomics Single Cell Feature_Barcode Library".
- __library step__ - Added "10x Genomics Single Cell V-D-J Library".
- __library step__ - Added ability for the lims to generate a liquid handling robot program to minimize the assigment of the same kapa tag twice in the same project. Both PCR-free and PCR-enriched library steps will have this feature.
- __sample qc__ - Adding steps for manual image attachment for Femto Pulse and Tapestation for DNA.
- __tube carrier__ - Adding carrier coordinate to DNA QC, PCR-free, PCR-enriched and WGBS protocols.
### Sample Manifest
- Moved the obligatory field marker (*) into the header itself, added more instructions in the manifest.
- Changed the color of the non-obligatory fields
- Changed "Sample in Pool" to "Library in Pool" for submitted library pools.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.3.0%0D0.2.10#diff)

## [0.2.10] - 2020-03-10
### LIMS Configuration
- __sample reception__, __all index validation__ - Adding MNaseSeq to library_protocol_list.csv for sample reception.
### Run Processing
- See above for library_protocol_list.csv.
- Adding Pierre Lepage to pre-validation run report emails.
- Remove it.genome from run monitor UPDATE but keep it on run monitor ALERT.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.10%0D0.2.9#diff)

## [0.2.9] - 2020-03-04
### LIMS Configuration
- __sample reception__, __all index validation__ - correction (no. 2) of nextera dual index plate.
- __sample reception__, __all index validation__ - adding index 10X Genomics Vision RNA Single Cell index and library type.
- __sample reception__, __all index validation__ - adding index 10x single cell atac seq adapters sequences.
- __sample reception__, __all index validation__ - adding index E5xx—Nextera Rapid Capture Custom Enrichment Kit.
- __sample reception__, __all index validation__ - adding index H5xx—Nextera DNA CD Indexes (combinatorialdual).
- __sample reception__, __all index validation__ - adding index N5xx—Nextera Index Kit.
- __sample reception__, __all index validation__ - adding index S5xx—Nextera XT Index Kit v2.
- __sample reception__, __all index validation__ - adding index for TruSeq Small RNA.
- __sample reception__, __all index validation__ - adding TruSeq Amplicon Kits.
- __sample reception__, __all index validation__ - adding custom D701-D704 in reverse complement.
### Run Processing
- See above adapter definitions
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.9%0D0.2.8#diff)

## [0.2.8] - 2020-02-17
### LIMS Configuration
- __index definition__ - Index definition correction for Nextera Dual index plate.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.8%0D0.2.7#diff)

## [0.2.7] - 2020-02-14
### LIMS Configuration
- __sequencing__ - Enabling autopooling for NovaSeq, HiSeqX, HiSeq4000.
- __sequencing__ - Display workflow name in all sequencer queues, make default group by project.
- __sequencing__ - Enabling autostart sequencing step for all sequencers.
- __sequencing__ - Disable autostart striptube step for HiSeqX and HiSeq4000.
- __sequencing__ - Disable autostart Clustering step for HiSeqX, HiSeq4000 and NovaSeq.
- __sequencing__ - Automatically create the runprocessing event file (hides button) when entering Sequencing steps for NovaSeq, HiSeqX, HiSeq4000, MiSeq, iSeq.
- __sequencing__ - Turn off lane fraction calculation as it overwrites what the user sets, affects NovaSeq, HiSeqX, HiSeq4000, MiSeq, iSeq.
- __sequencing__ - Changed MiSeq path to point to the correct Miseq drop folder for interop metrics script. Changing Technology Hash Key for metadata attachement.
- __sequencing__ - Removing striptube step for iSeq, MiSeq and NovaSeq.
- __sequencing__ - Reorder sequencing cycles to R1 I1 I2 R2 for all sequencers.
- __sequencing__ - Reorder Pooling Group(1), Lane Fraction(2), Index Cycles(3) for all sequencers.
### Run Processing
- Added rrna index path for Mus_musculus.GRCm38.
- Changed MiSeq path to point to the correct Miseq drop folder.
- Changed the "from" of the sequencing network copy alert email to bravo.genome@mcgill.ca.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.7%0D0.2.6#diff)

## [0.2.6] - 2020-01-27
### LIMS Configuration
- __sample reception__, __all index validation__ - Added NEBNext adapter definition.
### Run Processing
- Added NEBNext adapter definition.
### Sample Manifest
- Added NEBNext adapters to the dropdowns.
- Added Nextera ATAC Series adapters to the dropdowns.
- Added Omixon adapters to the dropdowns.
- Added Haloplex adapters to the dropdowns.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.6%0D0.2.5#diff)

## [0.2.5] - 2020-01-15
### LIMS Configuration
- __sample reception__, __all index validation__ - Added Omixon adapter definition.
### Run Processing
- Added Omixon adapter definition.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.5%0D0.2.4#diff)

## [0.2.4] - 2020-01-15
### LIMS Configuration
- __sequencing__, __library pooling__ - Harmonization with HiSeq X,4000 Novaseq and Pooling for captures.
All library pooling steps now have hamming distance based pooling ability based on sequence and not on name only (clarity default).
The change allows pooling of artifacts with apparent conflicting Reagent Label name for when we have an external libraries or when the index was corrected at the library level but not downstream.
This change affects MiSeq, iSeq and out of sequencing Library pooling only.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.4%0D0.2.3#diff)

## [0.2.3] - 2020-01-10
### Repository
- Added This change log file
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.3%0D0.2.2#diff)

## [0.2.2] - 2020-01-10
### LIMS Configuration
- __sample reception__ - Increased the sample submission sample name length limit to 60 from 40.
### Sample Manifest
- Increased the sample submission sample name length to 60 from 40 in the notes.
- Included basic sample name error detection in the excel file (length and illegal characters).
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.2%0D0.2.1#diff)

## [0.2.1] - 2020-01-09
### LIMS Configuration
- __library pooling__, __captures__ - Transferred the script with the ability to perform index compatibility for pooling for external libraries brought to Captures or Library Pooling.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.1%0D0.2.0#diff)

## [0.2.0] - 2020-01-07
### LIMS Configuration
- __Sequencing__ - Index validation for pooling for all sequencer types (libraries and library pools and captures, internal and external library or library pool).
- __Sequencing__ - MiSeq and HiSeq4000 worflows.
- __Sequencing__ - Output ChIP-Seq mark for run processing event file.
- __Sequencing__ - Handling of old library type "Lucigen AmpFree" and "Kapa Hyper Plus" for run processing event file.
- __Sequencing__ - Handle the case where the same library is sequenced in lowpass in one lane and inside a capture in another lane for the same flowcell.
- __Sample Reception__ - Only alphanumeric, integers underscore and dash being allowed, the lims will now not autocorrect but raise an alarm.
- __Sample Reception__ - The sample name length will be limited to 40.
- __Sample Reception__ - Library index for submitted libaries are verified.
### Run Processing
- Added ability to run the haloplex pipeline with 8 index cycles as well as 10.
- Increased wall time for fastqc_babraham to handle large fastq (>1G reads).
- Increased memory for readsqc to handle large fastq (>1G reads).
- Fixed an issue with bcl2fastq masks where only read1 is read.
- Removed False flag in the report when Species is not defined.
### Sample Manifest
- Edit the sample manifest to indicate that the sample name length should be less or equal to 40.
###
[see changes](https://bitbucket.org/mugqic/clarity/branches/compare/0.2.0%0D0.1.7#diff)

