ont_filelistpersample="alignment/PLACEHOLDER/PLACEHOLDER.sorted.bam
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.primerTrim.bam
artic_nanopolish/PLACEHOLDER/PLACEHOLDER.consensus.fasta
artic_nanopolish/PLACEHOLDER/PLACEHOLDER.muscle.in.fasta
artic_nanopolish/PLACEHOLDER/PLACEHOLDER.muscle.out.fasta
artic_nanopolish/PLACEHOLDER/PLACEHOLDER.pass.vcf.gz
artic_nanopolish/PLACEHOLDER/PLACEHOLDER.preconsensus.fasta
artic_nanopolish/PLACEHOLDER/PLACEHOLDER.primertrimmed.rg.sorted.bam
artic_nanopolish/PLACEHOLDER/PLACEHOLDER.sorted.bam
artic_nanopolish/PLACEHOLDER/PLACEHOLDER.trimmed.rg.sorted.bam
consensus/PLACEHOLDER/PLACEHOLDER.consensus.fasta
consensus/PLACEHOLDER/PLACEHOLDER.consensus.nanopore.pass.fasta
host_removal/PLACEHOLDER/PLACEHOLDER.host_removed.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.host_removed.sorted.bam
host_removal/PLACEHOLDER/PLACEHOLDER.hybrid.sorted.bam
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.classified_sequences.fastq.gz
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.kraken2_report
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.unclassified_sequences.fastq.gz
metrics/dna/PLACEHOLDER/quast_metrics/basic_stats/PLACEHOLDER.consensus_GC_content_plot.pdf
metrics/dna/PLACEHOLDER/quast_metrics/contigs_reports/all_alignments_PLACEHOLDER-consensus.tsv
variant/PLACEHOLDER/PLACEHOLDER.pass.annotate.vcf.gz
variant/PLACEHOLDER/PLACEHOLDER.pass.vcf.gz
report/ncov_tools/data/PLACEHOLDER.consensus.fasta
report/ncov_tools/data/PLACEHOLDER.pass.vcf.gz
report/ncov_tools/data/PLACEHOLDER.sorted.bam
report/ncov_tools/data/PLACEHOLDER.sorted.filtered.primerTrim.bam
report/ncov_tools/qc_analysis/PLACEHOLDER.summary.qc.tsv
report/ncov_tools/qc_annotation/PLACEHOLDER_aa_table.tsv
report/ncov_tools/qc_annotation/PLACEHOLDER.primer_snp_depth.tsv
report/ncov_tools/qc_annotation/PLACEHOLDER.primer_snps.bed
report/ncov_tools/qc_sequencing/PLACEHOLDER.amplicon_coverage.bed
report/ncov_tools/qc_sequencing/PLACEHOLDER.amplicon_depth.bed
report/ncov_tools/qc_sequencing/PLACEHOLDER.per_base_coverage.bed"

ont_filelistperrun="bam_alignment_qc.pdf
report/ncov_tools/bed/amplicon.bed
report/ncov_tools/bed/genome.bed
report/ncov_tools/metadata.tsv
report/ncov_tools/plots/*_aa_mutation_heatmap.pdf
report/ncov_tools/plots/*_amplicon_coverage_heatmap.pdf
report/ncov_tools/plots/*_amplicon_covered_fraction.pdf
report/ncov_tools/plots/*_depth_by_position.pdf
report/ncov_tools/plots/*_tree_snps.pdf
report/ncov_tools/qc_analysis/*_aligned.fasta
report/ncov_tools/qc_analysis/*_alleles.tsv
report/ncov_tools/qc_analysis/*_amplicon_coverage_table.tsv
report/ncov_tools/qc_analysis/*_consensus.fasta
report/ncov_tools/qc_annotation/*.primer_snp_depth_merged.tsv
report/ncov_tools/qc_annotation/*.primer_snp_depth_summary.tsv
report/ncov_tools/qc_reports/*_ncov_watch_variants.tsv
report/ncov_tools/qc_reports/*_summary_qc.tsv
report.readset.tsv"

illum_filelistpersample="alignment/PLACEHOLDER/PLACEHOLDER.*.1/PLACEHOLDER.*.1.sorted.bam
alignment/PLACEHOLDER/PLACEHOLDER.*.2/PLACEHOLDER.*.2.sorted.bam
alignment/PLACEHOLDER/PLACEHOLDER.primerTrim.bam
alignment/PLACEHOLDER/PLACEHOLDER.sorted.bam
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.all.metrics.base_distribution_by_cycle.pdf
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.all.metrics.insert_size_histogram.pdf
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.all.metrics.quality_by_cycle.pdf
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.all.metrics.quality_distribution.pdf
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.bam
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.coverage.tsv
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.onTarget.tsv
alignment/PLACEHOLDER/PLACEHOLDER.sorted.filtered.primerTrim.bam
cleaned_raw_reads/PLACEHOLDER/PLACEHOLDER.*.1.trim.pair1.fastq.gz
cleaned_raw_reads/PLACEHOLDER/PLACEHOLDER.*.1.trim.pair2.fastq.gz
cleaned_raw_reads/PLACEHOLDER/PLACEHOLDER.*.2.trim.pair1.fastq.gz
cleaned_raw_reads/PLACEHOLDER/PLACEHOLDER.*.2.trim.pair2.fastq.gz
consensus/PLACEHOLDER/PLACEHOLDER.consensus.fasta
consensus/PLACEHOLDER/PLACEHOLDER.consensus.illumina.pass.fasta
consensus/PLACEHOLDER/PLACEHOLDER.freebayes_calling.ambiguous.fasta
consensus/PLACEHOLDER/PLACEHOLDER.freebayes_calling.consensus.fasta
consensus/PLACEHOLDER/PLACEHOLDER.freebayes_calling.consensus.illumina.pass.fasta
consensus/PLACEHOLDER/PLACEHOLDER.freebayes_calling.consensus.renamed.fasta
host_removal/PLACEHOLDER/PLACEHOLDER.*.1.host_removed.nsorted.bam
host_removal/PLACEHOLDER/PLACEHOLDER.*.1.host_removed.other.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.*.1.host_removed.pair1.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.*.1.host_removed.pair2.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.*.1.host_removed.single.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.*.1.host_removed.sorted.bam
host_removal/PLACEHOLDER/PLACEHOLDER.*.1.hybrid.sorted.bam
host_removal/PLACEHOLDER/PLACEHOLDER.*.2.host_removed.nsorted.bam
host_removal/PLACEHOLDER/PLACEHOLDER.*.2.host_removed.other.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.*.2.host_removed.pair1.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.*.2.host_removed.pair2.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.*.2.host_removed.single.fastq.gz
host_removal/PLACEHOLDER/PLACEHOLDER.*.2.host_removed.sorted.bam
host_removal/PLACEHOLDER/PLACEHOLDER.*.2.hybrid.sorted.bam
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.1.classified_sequences_1.fastq.gz
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.1.classified_sequences_2.fastq.gz
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.1.kraken2_report
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.1.unclassified_sequences_1.fastq.gz
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.1.unclassified_sequences_2.fastq.gz
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.2.classified_sequences_1.fastq.gz
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.2.classified_sequences_2.fastq.gz
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.2.kraken2_report
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.2.unclassified_sequences_1.fastq.gz
metrics/dna/PLACEHOLDER/kraken_metrics/PLACEHOLDER.*.2.unclassified_sequences_2.fastq.gz
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.all.metrics.base_distribution_by_cycle.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.all.metrics.insert_size_histogram.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.all.metrics.quality_by_cycle.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.all.metrics.quality_distribution.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.filtered.all.metrics.base_distribution_by_cycle.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.filtered.all.metrics.insert_size_histogram.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.filtered.all.metrics.quality_by_cycle.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.filtered.all.metrics.quality_distribution.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.filtered.qcbias_metrics.pdf
metrics/dna/PLACEHOLDER/picard_metrics/PLACEHOLDER.sorted.qcbias_metrics.pdf
metrics/dna/PLACEHOLDER/quast_metrics_freebayes/basic_stats/PLACEHOLDER.freebayes_calling.consensus_GC_content_plot.pdf
metrics/dna/PLACEHOLDER/quast_metrics_freebayes/contigs_reports/all_alignments_PLACEHOLDER-freebayes_calling-consensus.tsv
metrics/dna/PLACEHOLDER/quast_metrics_ivar/basic_stats/PLACEHOLDER.sorted.filtered.primerTrim.consensus_GC_content_plot.pdf
metrics/dna/PLACEHOLDER/quast_metrics_ivar/contigs_reports/all_alignments_PLACEHOLDER-sorted-filtered-primerTrim-consensus.tsv
variant/PLACEHOLDER/PLACEHOLDER.freebayes_calling.ambiguous.norm.vcf.gz
variant/PLACEHOLDER/PLACEHOLDER.freebayes_calling.consensus.norm.vcf.gz
variant/PLACEHOLDER/PLACEHOLDER.freebayes_calling.fixed.norm.annotate.vcf.gz
variant/PLACEHOLDER/PLACEHOLDER.freebayes_calling.fixed.norm.vcf.gz
variant/PLACEHOLDER/PLACEHOLDER.freebayes_calling.variants.norm.vcf.gz
variant/PLACEHOLDER/PLACEHOLDER.sorted.filtered.primerTrim.annotate.vcf.gz
variant/PLACEHOLDER/PLACEHOLDER.sorted.filtered.primerTrim.vcf.gz
variant/PLACEHOLDER/PLACEHOLDER.variants.tsv
report/ncov_tools_freebayes/data/PLACEHOLDER.freebayes_calling.consensus.renamed.fasta
report/ncov_tools_freebayes/data/PLACEHOLDER.sorted.filtered.bam
report/ncov_tools_freebayes/data/PLACEHOLDER.sorted.filtered.primerTrim.bam
report/ncov_tools_freebayes/qc_analysis/PLACEHOLDER.summary.qc.tsv
report/ncov_tools_freebayes/qc_annotation/PLACEHOLDER_aa_table.tsv
report/ncov_tools_freebayes/qc_annotation/PLACEHOLDER.primer_snp_depth.tsv
report/ncov_tools_freebayes/qc_annotation/PLACEHOLDER.primer_snps.bed
report/ncov_tools_freebayes/qc_sequencing/PLACEHOLDER.amplicon_coverage.bed
report/ncov_tools_freebayes/qc_sequencing/PLACEHOLDER.amplicon_depth.bed
report/ncov_tools_freebayes/qc_sequencing/PLACEHOLDER.fpileup.tsv
report/ncov_tools_freebayes/qc_sequencing/PLACEHOLDER.per_base_coverage.bed
report/ncov_tools_freebayes/tmp_bam/PLACEHOLDER.bam
report/ncov_tools_ivar/data/PLACEHOLDER.consensus.fasta
report/ncov_tools_ivar/data/PLACEHOLDER.sorted.filtered.bam
report/ncov_tools_ivar/data/PLACEHOLDER.sorted.filtered.primerTrim.bam
report/ncov_tools_ivar/data/PLACEHOLDER.variants.tsv
report/ncov_tools_ivar/qc_analysis/PLACEHOLDER.summary.qc.tsv
report/ncov_tools_ivar/qc_annotation/PLACEHOLDER_aa_table.tsv
report/ncov_tools_ivar/qc_annotation/PLACEHOLDER.pass.vcf.gz
report/ncov_tools_ivar/qc_annotation/PLACEHOLDER.primer_snp_depth.tsv
report/ncov_tools_ivar/qc_annotation/PLACEHOLDER.primer_snps.bed
report/ncov_tools_ivar/qc_sequencing/PLACEHOLDER.amplicon_coverage.bed
report/ncov_tools_ivar/qc_sequencing/PLACEHOLDER.amplicon_depth.bed
report/ncov_tools_ivar/qc_sequencing/PLACEHOLDER.fpileup.tsv
report/ncov_tools_ivar/qc_sequencing/PLACEHOLDER.per_base_coverage.bed
report/ncov_tools_ivar/tmp_bam/PLACEHOLDER.bam"

illum_filelistperrun="metrics/host_contamination_metrics.tsv
metrics/host_removed_metrics.tsv
metrics/kraken2_metrics.tsv
report/ncov_tools_freebayes/bed/amplicon.bed
report/ncov_tools_freebayes/bed/genome.bed
report/ncov_tools_freebayes/metadata.tsv
report/ncov_tools_freebayes/plots/*_aa_mutation_heatmap.pdf
report/ncov_tools_freebayes/plots/*_amplicon_coverage_heatmap.pdf
report/ncov_tools_freebayes/plots/*_amplicon_covered_fraction.pdf
report/ncov_tools_freebayes/plots/*_depth_by_position.pdf
report/ncov_tools_freebayes/plots/*_tree_snps.pdf
report/ncov_tools_freebayes/qc_analysis/*_aligned.fasta
report/ncov_tools_freebayes/qc_analysis/*_alleles.tsv
report/ncov_tools_freebayes/qc_analysis/*_amplicon_coverage_table.tsv
report/ncov_tools_freebayes/qc_analysis/*_consensus.fasta
report/ncov_tools_freebayes/qc_annotation/*.primer_snp_depth_merged.tsv
report/ncov_tools_freebayes/qc_annotation/*.primer_snp_depth_summary.tsv
report/ncov_tools_freebayes/qc_reports/*_ambiguous_position_report.tsv
report/ncov_tools_freebayes/qc_reports/*_ncov_watch_variants.tsv
report/ncov_tools_freebayes/qc_reports/*_negative_control_report.tsv
report/ncov_tools_freebayes/qc_reports/*_summary_qc.tsv
report/ncov_tools_ivar/bed/amplicon.bed
report/ncov_tools_ivar/bed/genome.bed
report/ncov_tools_ivar/metadata.tsv
report/ncov_tools_ivar/plots/*_aa_mutation_heatmap.pdf
report/ncov_tools_ivar/plots/*_amplicon_coverage_heatmap.pdf
report/ncov_tools_ivar/plots/*_amplicon_covered_fraction.pdf
report/ncov_tools_ivar/plots/*_depth_by_position.pdf
report/ncov_tools_ivar/plots/*_tree_snps.pdf
report/ncov_tools_ivar/qc_analysis/*_aligned.fasta
report/ncov_tools_ivar/qc_analysis/*_alleles.tsv
report/ncov_tools_ivar/qc_analysis/*_amplicon_coverage_table.tsv
report/ncov_tools_ivar/qc_analysis/*_consensus.fasta
report/ncov_tools_ivar/qc_annotation/*.primer_snp_depth_merged.tsv
report/ncov_tools_ivar/qc_annotation/*.primer_snp_depth_summary.tsv
report/ncov_tools_ivar/qc_reports/*_ambiguous_position_report.tsv
report/ncov_tools_ivar/qc_reports/*_ncov_watch_variants.tsv
report/ncov_tools_ivar/qc_reports/*_negative_control_report.tsv
report/ncov_tools_ivar/qc_reports/*_summary_qc.tsv
report.readset_freebayes.tsv
report.readset_ivar.tsv
report/report_metrics_freebayes.tsv
report/report_metrics_ivar.tsv"

